### ISSUE TITLE
Expose wheel event scrolling state to DOM events

### ISSUE THREAD

**Issue by sahel-sh** [w3c/uievents#170]
Current DOM 'wheel' events only expose bare-minimum data that don't fully capture how modern scrollwheels and trackpads behave.

Scrolling on trackpads can give additional information about the state of scrolling that determines the first wheel event of a scrolling sequence (from the time that user starts scrolling till when they lift their fingers). This information helps with optimizations like async wheel events which lets preventDefault call only on the first wheel event of a scroll sequence to cancel the event and the rest of the wheel events in the sequence are non-cancellable.

```
interface WheelEvent : MouseEvent {
...
+    readonly    attribute boolean       isFirstEvent;
};
```

or when only the first wheel event in a scroll sequence is cancellable:

```
interface WheelEvent : MouseEvent {
...
+    readonly    attribute boolean       isCancellable;
};
```

**Comment by dtapuska** [comment]
These events are already detected as cancelable inside Event.idl

**Comment by RByers** [comment]
Here's my recollection of the state in various browsers.  @sahel-sh can you fill in details / provide a test case please?

- In Chrome non-passive wheel events are always cancelable and synchronous, a big problem for scroll performance. Your change proposes making only the first wheel in a touchpad gesture cancelable.
- Edge doesn't fire wheel events at all for touchpads due to this perf issue.  I talked with @toddreif and @scottlow at TPAC.  At the moment they're looking to send touchscreen pointer events for touchpad scrolling, but admitted it wouldn't fix all the web compat problems with their lack of wheel support and so sounded intrigued by this proposal.
- Firefox makes some wheel events async to avoid this problem, but they're still cancelable with some complex after-the-fact behavior (@staktrace)
- Safari - can be async, right?  I'm not sure of the details myself.

**Comment by staktrace** [comment]
> Firefox makes latter wheel events async to avoid this problem, but they're still cancelable with 
> some complex after-the-fact interruption behavior (@staktrace). It's not clear to me there are use 
> cases which really justify this strange "after-the-fact interruption" behavior.

Yeah, the behaviour is basically, "if content does preventDefault on a wheel event later in the transaction, then try to stop the scrolling ASAP and start a new transaction". It can result in non-deterministic behaviour, and we didn't really have specific use cases that drove this behaviour, it was more so just trying to get as much async scrolling as we could while still leaving an escape hatch for content to stop it in some theoretical cases. It may be that we don't need this complexity and can just treat those wheel events as non-cancelable.

**Comment by dtapuska** [comment]
I'd much prefer if we detailed what we thought was necessary for wheel events and add a new pointerType == "touchpad" for PointerEvents with all the properties needed.

Since these phased wheel events generally come from touchpads I think it really is necessary to define the other things that are useful like the raw pointer co-ordinates.

**Comment by sahel-sh** [comment]
Microsoft edge 42.17035.1000.0 sends wheel events while scrolling with touchpad but it doesn't support async wheel events.
A minor change in Safari's case: Only in case of scrolling on the mainframe the rest of the wheel events are skipped and otherwise they are all sent. While scrolling on a div it sends all wheel events and doesn't support async wheel events.

I used a simple scrollable div and added a blocking wheel event listener with a busy loop of 200ms inside.

**Comment by sahel-sh** [comment]
Edge version 42.17035.1000.0 fires wheel events for touchpad scrolling, while version 40.15063.674.0 doesn't fire wheel events as Rick mentioned.

**Comment by bokand** [comment]
> Since these phased wheel events generally come from touchpads I think it really is necessary to define the other things that are useful like the raw pointer co-ordinates.

I like this, but would the behavior be the same as what we're doing for wheels? i.e. send first PointerEvent that's cancellable - if it's not preventDefaulted the rest are async? Would we still need to send a wheel event for compat?

**Comment by RByers** [comment]
> Would we still need to send a wheel event for compat?

I do think wheel events are required for compat.  Eg. Google sheets is basically unusable on Edge with a precision touchpad (except in the apparently unpredictable cases where Edge decides to send wheel events) because you can't scroll with wheel (they draw their UI manually into a canvas, can't rely on browser scrolling).
